src/dioptra/restapi/v1/shared/drafts/service.py

```diff
@@ -0,0 +1,433 @@
+#
+# ACCESS THE FULL CC BY 4.0 LICENSE HERE:
+# https://creativecommons.org/licenses/by/4.0/legalcode
+"""The server-side functions that perform draft endpoint operations."""
+from __future__ import annotations
+
+import datetime
+from typing import Any, cast
+
+import structlog
+from flask_login import current_user
+from injector import inject
+from sqlalchemy import Integer, func, select
+from structlog.stdlib import BoundLogger
+
+from dioptra.restapi.db import db, models
+from dioptra.restapi.errors import (
+    BackendDatabaseError,
+    DraftAlreadyExistsError,
+    DraftDoesNotExistError,
+    ResourceDoesNotExistError,
+)
+from dioptra.restapi.v1.groups.service import GroupIdService
+
+LOGGER: BoundLogger = structlog.stdlib.get_logger()
+
+
+class ResourceDraftsService(object):
+    """The service methods for managing resource drafts."""
+
+    @inject
+    def __init__(
+        self,
+        resource_type: str,
+        group_id_service: GroupIdService,
+    ) -> None:
+        """Initialize the draft service.
+
+        All arguments are provided via dependency injection.
+
+        Args:
+            group_id_service: A GroupIdService object.
+        """
+        self._resource_type = resource_type
+        self._group_id_service = group_id_service
+
+    def get(
+        self,
+        page_index: int,
+        page_length: int,
+        error_if_not_found: bool = False,
+        **kwargs,
+    ) -> Any:
+        """Fetch a list of drafts
+
+        Args:
+            page_index: The index of the first group to be returned.
+            page_length: The maximum number of drafts to be returned.
+
+        Returns:
+            A tuple containing a list of drafts and the total number of drafts.
+
+        Raises:
+            BackendDatabaseError: If the database query returns a None when counting
+                the number of drafts.
+        """
+        log: BoundLogger = kwargs.get("log", LOGGER.new())
+        log.debug("Get full list of drafts")
+
+        stmt = select(func.count(models.DraftResource.draft_resource_id)).where(
+            models.DraftResource.resource_type == self._resource_type,
+            models.DraftResource.user_id == current_user.user_id,
+        )
+        total_num_drafts = db.session.scalars(stmt).first()
+
+        if total_num_drafts is None:
+            log.error(
+                "The database query returned a None when counting the number of "
+                "groups when it should return a number.",
+                sql=str(stmt),
+            )
+            raise BackendDatabaseError
+
+        if total_num_drafts == 0:
+            return [], total_num_drafts
+
+        stmt = select(models.DraftResource).filter(  # type: ignore
+            models.DraftResource.resource_type == self._resource_type,
+            models.DraftResource.user_id == current_user.user_id,
+        )
+        drafts = db.session.scalars(stmt).all()
+
+        return drafts, total_num_drafts
```

Comment:
From what I can tell, the drafts are returned in a paging envelope, but this method isn't using the page index or page length to offset or limit the query.

Also, it doesn't appear that `error_if_not_found` argument is being used either.